### ISSUE TITLE
`conda install` or `conda create` still has trouble resolving features when specified

### ISSUE THREAD

**Issue by jakirkham** [conda/conda#1476]
Suppose I install the following program from my own channel. It does not grab the latest versions of `numpy`, `scipy`, and `spams` that I have. Note that none of the version are fixed in my package's recipe. Note: I am not sure if this is an issue with `conda` or with [anaconda.org](https://anaconda.org).

To reproduce the channel has changed to `nanshe`. Behavior appears to be the same though.

```
$ conda install -y -c jakirkham nanshe
Fetching package metadata: ......
Solving package specifications: .....................
Package plan for installation in environment /opt/conda:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    fftw-3.3.4                 |                0         3.5 MB
    gmp-5.1.2                  |                2         594 KB
    icu-54.1                   |                0        11.3 MB
    pixman-0.26.2              |                0         1.9 MB
    boost-1.57.0               |                4         9.7 MB
    hdf5-1.8.15.1              |                1         1.8 MB
    isl-0.12.2                 |                0         1.1 MB
    libxml2-2.9.2              |                0         4.2 MB
    mpfr-3.1.2                 |                0         407 KB
    cloog-0.18.0               |                0         617 KB
    fontconfig-2.11.1          |                4         402 KB
    mpc-1.0.1                  |                0          61 KB
    cairo-1.12.18              |                4         592 KB
    libgcc-4.8.4               |                1         3.6 MB
    qt-4.8.6                   |                3        36.3 MB
    openblas-0.2.14            |              672         6.7 MB
    decorator-4.0.2            |           py27_0          11 KB
    futures-3.0.3              |           py27_0          17 KB
    networkx-1.9.1             |           py27_0         934 KB
    nose-1.3.7                 |           py27_0         194 KB
    numpy-1.9.2                |           py27_2         6.2 MB
    bottleneck-1.0.0           |       np19py27_0         533 KB
    psutil-2.2.1               |           py27_0          98 KB
    py2cairo-1.10.0            |           py27_2          36 KB
    pyfftw-0.9.2               |       np19py27_1         199 KB
    pyparsing-2.0.3            |           py27_0          63 KB
    scipy-0.15.1               |       np19py27_2        28.2 MB
    sip-4.16.5                 |           py27_0         245 KB
    vigra-ge4946fe             |       np19py27_3         6.4 MB
    h5py-2.5.0                 |       np19py27_3         2.3 MB
    pyqt-4.11.3                |           py27_1         3.5 MB
    rank_filter-v0.1           |       np19py27_0         264 KB
    spams-v2.4                 |       np19py27_0         4.2 MB
    matplotlib-1.4.3           |       np19py27_2        44.5 MB
    qimage2ndarray-1.4         |       np19py27_0          20 KB
    scikit-image-0.11.3        |       np19py27_0        23.8 MB
    volumina-v1.1.6a7          |       np19py27_0         1.0 MB
    nanshe-0.1.0a22            |       np19py27_0         204 KB
    ------------------------------------------------------------
                                           Total:       205.8 MB

The following NEW packages will be INSTALLED:

    boost:          1.57.0-4           
    bottleneck:     1.0.0-np19py27_0   
    cairo:          1.12.18-4          
    cloog:          0.18.0-0           
    decorator:      4.0.2-py27_0       
    fftw:           3.3.4-0            
    fontconfig:     2.11.1-4           
    futures:        3.0.3-py27_0       
    gmp:            5.1.2-2            
    h5py:           2.5.0-np19py27_3   
    hdf5:           1.8.15.1-1         
    icu:            54.1-0             
    isl:            0.12.2-0           
    libgcc:         4.8.4-1            
    libxml2:        2.9.2-0            
    matplotlib:     1.4.3-np19py27_2   
    mpc:            1.0.1-0            
    mpfr:           3.1.2-0            
    nanshe:         0.1.0a22-np19py27_0
    networkx:       1.9.1-py27_0       
    nose:           1.3.7-py27_0       
    numpy:          1.9.2-py27_2       
    openblas:       0.2.14-672         
    pixman:         0.26.2-0           
    psutil:         2.2.1-py27_0       
    py2cairo:       1.10.0-py27_2      
    pyfftw:         0.9.2-np19py27_1   
    pyparsing:      2.0.3-py27_0       
    pyqt:           4.11.3-py27_1      
    qimage2ndarray: 1.4-np19py27_0     
    qt:             4.8.6-3            
    rank_filter:    v0.1-np19py27_0    
    scikit-image:   0.11.3-np19py27_0  
    scipy:          0.15.1-np19py27_2  
    sip:            4.16.5-py27_0      
    spams:          v2.4-np19py27_0    
    vigra:          ge4946fe-np19py27_3
    volumina:       v1.1.6a7-np19py27_0

Fetching packages ...
fftw-3.3.4-0.t 100% |################################################################################################################################| Time: 0:00:00   6.05 MB/s
gmp-5.1.2-2.ta 100% |################################################################################################################################| Time: 0:00:00   2.09 MB/s
icu-54.1-0.tar 100% |################################################################################################################################| Time: 0:00:01   7.32 MB/s
pixman-0.26.2- 100% |################################################################################################################################| Time: 0:00:00   3.52 MB/s
boost-1.57.0-4 100% |################################################################################################################################| Time: 0:00:01   6.49 MB/s
hdf5-1.8.15.1- 100% |################################################################################################################################| Time: 0:00:00   3.25 MB/s
isl-0.12.2-0.t 100% |################################################################################################################################| Time: 0:00:00   3.22 MB/s
libxml2-2.9.2- 100% |################################################################################################################################| Time: 0:00:01   4.28 MB/s
mpfr-3.1.2-0.t 100% |################################################################################################################################| Time: 0:00:00   1.96 MB/s
cloog-0.18.0-0 100% |################################################################################################################################| Time: 0:00:00   2.31 MB/s
fontconfig-2.1 100% |################################################################################################################################| Time: 0:00:00   1.69 MB/s
mpc-1.0.1-0.ta 100% |################################################################################################################################| Time: 0:00:00   1.25 MB/s
cairo-1.12.18- 100% |################################################################################################################################| Time: 0:00:00   2.04 MB/s
libgcc-4.8.4-1 100% |################################################################################################################################| Time: 0:00:00   5.02 MB/s
qt-4.8.6-3.tar 100% |################################################################################################################################| Time: 0:00:03  11.13 MB/s
openblas-0.2.1 100% |################################################################################################################################| Time: 0:00:01   6.29 MB/s
decorator-4.0. 100% |################################################################################################################################| Time: 0:00:00   5.18 MB/s
futures-3.0.3- 100% |################################################################################################################################| Time: 0:00:00   1.15 MB/s
networkx-1.9.1 100% |################################################################################################################################| Time: 0:00:00   2.61 MB/s
nose-1.3.7-py2 100% |################################################################################################################################| Time: 0:00:00   1.72 MB/s
numpy-1.9.2-py 100% |################################################################################################################################| Time: 0:00:00   7.35 MB/s
bottleneck-1.0 100% |################################################################################################################################| Time: 0:00:00   2.03 MB/s
psutil-2.2.1-p 100% |################################################################################################################################| Time: 0:00:00 813.02 kB/s
py2cairo-1.10. 100% |################################################################################################################################| Time: 0:00:00 967.41 kB/s
pyfftw-0.9.2-n 100% |################################################################################################################################| Time: 0:00:00   2.10 MB/s
pyparsing-2.0. 100% |################################################################################################################################| Time: 0:00:00 958.18 kB/s
scipy-0.15.1-n 100% |################################################################################################################################| Time: 0:00:02  10.74 MB/s
sip-4.16.5-py2 100% |################################################################################################################################| Time: 0:00:00 940.99 kB/s
vigra-ge4946fe 100% |################################################################################################################################| Time: 0:00:01   6.29 MB/s
h5py-2.5.0-np1 100% |################################################################################################################################| Time: 0:00:01   2.05 MB/s
pyqt-4.11.3-py 100% |################################################################################################################################| Time: 0:00:01   3.34 MB/s
rank_filter-v0 100% |################################################################################################################################| Time: 0:00:00   1.05 MB/s
spams-v2.4-np1 100% |################################################################################################################################| Time: 0:00:00   5.00 MB/s
matplotlib-1.4 100% |################################################################################################################################| Time: 0:00:04  10.76 MB/s
qimage2ndarray 100% |################################################################################################################################| Time: 0:00:00   1.32 MB/s
scikit-image-0 100% |################################################################################################################################| Time: 0:00:02   9.80 MB/s
volumina-v1.1. 100% |################################################################################################################################| Time: 0:00:00   3.08 MB/s
nanshe-0.1.0a2 100% |################################################################################################################################| Time: 0:00:00   2.00 MB/s
Extracting packages ...
[      COMPLETE      ]|###################################################################################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|###################################################################################################################################################| 100%
```

To demonstrate this issue I can update to this effect.

```
$ conda update -y -c jakirkham --all

Fetching package metadata: ......
Solving package specifications: .
Package plan for installation in environment /opt/conda:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    numpy-1.9.2                |           py27_3         6.2 MB
    scipy-0.16.0               |     np19py27_100        29.4 MB
    spams-v2.5                 |       np19py27_0         4.4 MB
    ------------------------------------------------------------
                                           Total:        40.0 MB

The following packages will be UPDATED:

    numpy: 1.9.2-py27_2      --> 1.9.2-py27_3        [openblas]
    scipy: 0.15.1-np19py27_2 --> 0.16.0-np19py27_100 [openblas]
    spams: v2.4-np19py27_0   --> v2.5-np19py27_0     [openblas]

Fetching packages ...
numpy-1.9.2-py 100% |################################################################################################################################| Time: 0:00:00   6.72 MB/s
scipy-0.16.0-n 100% |################################################################################################################################| Time: 0:00:02  11.13 MB/s
spams-v2.5-np1 100% |################################################################################################################################| Time: 0:00:00   5.66 MB/s
Extracting packages ...
[      COMPLETE      ]|###################################################################################################################################################| 100%
Unlinking packages ...
[      COMPLETE      ]|###################################################################################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|###################################################################################################################################################| 100%
```

**Comment by jakirkham** [comment]
It turns out that the versions installed here are the last versions before feature tracking was enabled. In other words, I suspect the problem here is that feature tracking is not checked in dependency resolution.

**Comment by stuarteberg** [comment]
Probably related (or the same as): #1334, #1338, #1589.

**Comment by jakirkham** [comment]
I saw #1334 previously before when I hadn't update `scipy`. After updating `scipy`, I ran into some other much more irritating and less clear problems likely related to this one (e.g. unable to find dependency matching constraint). I wasn't 100% sure they were the same. It does appear that @ukoethe is describing a similar problem to mine in #1589.

**Comment by ukoethe** [comment]
According to my understanding, conda has a fast version resolution algorithm that sometimes (actually, often) misses features, and a correct one that is too slow by a factor of 10. @asmeurer is trying to speed-up the correct algorithm in order to phase-out the old code.

In my Windows builds, I work around this issue by appending `.vc11` to the version string whenever a package defines `features: -vc11`. Dependencies are specified as `foo *.vc11` so that non-feature versions are no longer considered in package resolution, even by the old algorithm. To make this nicer, I introduced jinja variables for these strings, see for example [hdf5/mata.yaml](https://github.com/ilastik/ilastik-build-conda/blob/toolset-config/hdf5/meta.yaml). This technique can readily be transformed to other matching problems. However, it depends on the pending PRs https://github.com/conda/conda/pull/1601 and https://github.com/conda/conda-build/pull/575.

**Comment by jakirkham** [comment]
Thanks for sharing @ukoethe. That seems like a cool solution.

At present, I have tried to mitigate this problem by building a copy of all packages at the latest version without the feature (or tracking of) and then building the packages with the same version and one higher build number with the feature. The net effect is the feature eventually gets added in once updated. It's using this bug, but it is at least going in my favor and won't hurt me once it is fixed.

Of course, this will break when a new version of one of these packages appears in conda. However, this is something I am willing to accept as it helps alert me to the fact that I should rebuild those packages to take advantage of the newer features and be more in line with the rest of the conda stack. Also, the packages that are tracking features are updated slowly so this isn't a common problem.

**Comment by asmeurer** [comment]
That workaround sounds like it should work.

**Comment by jakirkham** [comment]
@ukoethe, wanted to share this with you as you may find it of interest. ( https://github.com/conda/conda-recipes/pull/362#issuecomment-149197350 )

**Comment by jakirkham** [comment]
@msarahan, given the greater prevalence of features in packages I would like to discuss this with you at some point.

**Comment by msarahan** [comment]
Thanks for pinging me. I agree that this issue is pressing, but I must
defer to our new technical lead on this, @kalefranz for prioritization and
how to approach.

**Comment by kalefranz** [comment]
@jakirkham @msarahan Having an onsite chat with several people about solver issues tomorrow.  There's a lot to catch up on, but I'm definitely aware that the solver is a major pain point. @msarahan feel free to track me down tomorrow too and maybe you can catch me up on things here from your perspective.

**Comment by mcg1969** [comment]
You mentioned this issue is still open. Can you offer a way to reproduce, ideally some way to start with a bare environment (e.g., `conda create -n`...)

**Comment by jakirkham** [comment]
Yes, run `conda install -y -c nanshe -n nanshenv nanshe`.

**Comment by mcg1969** [comment]
Super, thanks

**Comment by jakirkham** [comment]
Sorry should have been `create`. The moral of the story is I have a dependency that depends on having packages with the feature only. I was unable to use Accelerate on Mac so had to use OpenBLAS.

**Comment by jakirkham** [comment]
There should be a simpler example that would do the same thing. This might be a place to start ( https://github.com/jakirkham/test_conda_features/ ).

**Comment by jakirkham** [comment]
This now also will produce the version/feature issue as there is a newer version of SciPy that doesn't track OpenBLAS. However, it is now tracking MKL so this may be an entirely different problem than it was before. Not really sure what the right solution is there.

**Comment by mcg1969** [comment]
I do see a problem with the code---exactly the reason I'm glad we put a pre-release out there.

**Comment by mcg1969** [comment]
OK, so I am investigating a fix. It comes in two parts: first is modifying [these lines](https://github.com/conda/conda/blob/master/conda/resolve.py#L340-L348) in resolve.py so that `prune_features` isn't called first, but second. If you're daring, here's the replacement:
```
        # Iterate in the filtering process until no more progress is made
        if filtered:
            pruned = True
            while pruned:
                for s in list(specs):
                    pruned = False
                    if not s.optional:
                        pruned += filter_group([s], None)
                        if unsat and sat_only:
                            return False
                    pruned += prune_features()
```
That causes the solver to pick up numpy and spasms properly, but it _doesn't_ pick up scipy. The reason for this, I think, is that there is no `scipy 0.17.0` available with the `openblas` feature, so it prefers to install the one without it than the older version with it.

Whether _that_ is a bug is probably up for discussion. My thinking is that yes, it is: if a feature is present in the environment, any package that has that among its options should use it, even if that means selecting an older version. @kalefranz, @groutr, what do you think? If we agree on this point, then I know how to fix it---it requires an additional `optimize()` call to maximize the total number of packages installed with `features` active.

**Comment by mcg1969** [comment]
Actually, I can do it without a new optimization pass. I just need to change the sort priority to look at the number of installed features before the version number.

**Comment by mcg1969** [comment]
If you want to hack on _that_, just change [this line](https://github.com/conda/conda/blob/master/conda/resolve.py#L445) from
```
        return (VersionOrder(rec['version']), len(feats), feats, rec['build_number'], rec['build'])
```
to
```
        return (len(feats), VersionOrder(rec['version']), rec['build_number'], rec['build'], feats)
```
Note this is not enough for a permanent fix, because of some subtleties about what happens when updating an installed package. But this will work fine for an installation.

**Comment by jakirkham** [comment]
Cool stuff, thanks for looking at this so quickly @mcg1969. I'll try these.

As for the scipy stuff, I think I agree with you that this is a bug. It should try to respect the feature. This is the most related issue ( https://github.com/conda/conda/issues/1334 ) I can think of regarding this. However, it is dealing with an upgrade where no feature is kept instead of the feature being change. Though I would think that difference is a minor point and most people would find both surprising, but I could be wrong.

**Comment by mcg1969** [comment]
One of the problems I'm seeing is that `openblas` sometimes provides the `openblas` track_feature, and sometimes it doesn't. Combine that with the fact that `openblas` is a _dependency_ of `numpy`, and I think you're going to see some weird behavior.

For instance, suppose I did this:
```
conda create -c nanshe -n test numpy=1.9.3 python=2.7*
```
Here's what I get:
```
    libgcc:     4.8.5-1       
    numpy:      1.9.3-py27_100
    openblas:   0.2.15-101    
    openssl:    1.0.2f-0      
    pip:        8.0.2-py27_0  
    python:     2.7.11-0      
    readline:   6.2-2         
    setuptools: 19.6.2-py27_0 
    sqlite:     3.9.2-0       
    tk:         8.5.18-0      
    wheel:      0.26.0-py27_1 
    zlib:       1.2.8-0       
```
Note that openblas is installed, and the version of numpy selected does depend on it, but the `openblas` feature is not selected. This looks right. On the other hand, if I do this:
```
conda create -c nanshe -n test3 numpy=1.9.3 python=2.7\* openblas=0.2.15=101
```
I get a different recipe:
```
    libgcc:     4.8.5-1       
    numpy:      1.9.3-py27_101 [openblas]
    openblas:   0.2.15-101    
    openssl:    1.0.2f-0      
    pip:        8.0.2-py27_0  
    python:     2.7.11-0      
    readline:   6.2-2         
    setuptools: 19.6.2-py27_0 
    sqlite:     3.9.2-0       
    tk:         8.5.18-0      
    wheel:      0.26.0-py27_1 
    zlib:       1.2.8-0 
```
So far, so good. Now: what *should* `conda` do with this recipe?
```
conda create -c nanshe -n test3 numpy=1.9.3 python=2.7\* openblas=0.2.15
```
My view is this: `conda` *should* give me the first recipe, but it is giving me the second, with the `openblas` feature enabled. Here's what I think conda's priorities should be:

_Priority 1_: minimize the number of installed features.
_Priority 2_: maximize the number of packages that utilize an installed feature.

I hope it is clear why Priority 1 must be true. If it were not, then `conda` would activate features just because they are in the repository, whether or not the user tried to select them. So it should always attempt to install the requested packages without adding features. So in this case, because `openblas` is available without the `track_feature` setting, `conda` should be attempting do without it.

Now of course, the more common ways of enabling a feature work great in this scenario:

1. If the `track_feature` is in a separate metapackage, then conda does the right thing, because installing that package means the feature _must_ be enabled.

2. If _every_ version of a package has the same `track_feature`, then again, if that package is installed by specification or by a strong dependency, conda does the right thing and enables it.

3. This works even if every version of a package has a `track_feature` but a _different_ one. For instance, our `python` packages have `vc9`, `vc10`, or `vc14` values, depending on the runtime. As long as they all have exactly one, `conda` will do the right thing in Priority 1 and select exactly one of them.

**Comment by mcg1969** [comment]
@kalefranz, @groutr, see my previous comment above.

**Comment by mcg1969** [comment]
OK, I've added the necessary optimization passes in `resolve.py`: first, minimize the feature count, and second, maximize the installed features. This is what I get for `conda create -n test -c nanshe nanshe`:
```
    boost:           1.57.0-4          
    bottleneck:      1.0.0-np110py27_0 
    cycler:          0.9.0-py27_0      
    decorator:       4.0.6-py27_0      
    fftw:            3.3.4-1           
    freetype:        2.5.5-0           
    functools32:     3.2.3.2-py27_0    
    futures:         3.0.3-py27_0      
    h5py:            2.5.0-np110py27_4 
    hdf5:            1.8.15.1-2        
    icu:             54.1-0            
    jbig:            2.1-0             
    jpeg:            8d-1              
    libgcc:          4.8.5-1           
    libpng:          1.6.17-0          
    libtiff:         4.0.6-1           
    mahotas:         1.4.0-py27_0      
    matplotlib:      1.5.1-np110py27_0 
    mkl:             11.3.1-0          
    nanshe:          0.1.0a44-py27_0   
    networkx:        1.11-py27_0       
    nose:            1.3.7-py27_0      
    numpy:           1.10.4-py27_102   
    openblas:        0.2.15-101        
    openssl:         1.0.2f-0          
    pillow:          3.1.0-py27_0      
    pip:             8.0.2-py27_0      
    psutil:          3.4.2-py27_0      
    pyfftw:          0.9.2-py27_2      
    pyparsing:       2.0.3-py27_0      
    pyqt:            4.11.4-py27_1     
    python:          2.7.11-0          
    python-dateutil: 2.4.2-py27_0      
    pytz:            2015.7-py27_0     
    qimage2ndarray:  1.4-py27_1        
    qt:              4.8.7-1           
    rank_filter:     0.4.6-nppy27      
    readline:        6.2-2             
    scikit-image:    0.11.3-np110py27_0
    scipy:           0.17.0-np110py27_0
    setuptools:      19.6.2-py27_0     
    sip:             4.16.9-py27_0     
    six:             1.10.0-py27_0     
    spams:           v2.5-py27_9       
    sqlite:          3.9.2-0           
    tk:              8.5.18-0          
    vigra:           gca22467-py27_4   
    volumina:        v1.1.8-py27_1     
    wheel:           0.26.0-py27_1     
    xz:              5.0.5-0           
    zlib:            1.2.8-0
```
Again, it chooses an `openblas` that does not have the `track_feature`. I can now force it with `conda create -n test3 -c nanshe nanshe openblas=0.2.15=101`:
```
    boost:           1.57.0-4                
    bottleneck:      1.0.0-np110py27_0       
    cycler:          0.9.0-py27_0            
    decorator:       4.0.6-py27_0            
    fftw:            3.3.4-1                 
    freetype:        2.5.5-0                 
    functools32:     3.2.3.2-py27_0          
    futures:         3.0.3-py27_0            
    h5py:            2.5.0-np110py27_4       
    hdf5:            1.8.15.1-2              
    icu:             54.1-0                  
    jbig:            2.1-0                   
    jpeg:            8d-1                    
    libgcc:          4.8.5-1                 
    libpng:          1.6.17-0                
    libtiff:         4.0.6-1                 
    mahotas:         1.4.0-py27_0            
    matplotlib:      1.5.1-np110py27_0       
    nanshe:          0.1.0a44-py27_0         
    networkx:        1.11-py27_0             
    nose:            1.3.7-py27_0            
    numpy:           1.10.4-py27_openblas_103 [openblas]
    openblas:        0.2.15-101              
    openssl:         1.0.2f-0                
    pillow:          3.1.0-py27_0            
    pip:             8.0.2-py27_0            
    psutil:          3.4.2-py27_0            
    pyfftw:          0.9.2-py27_2            
    pyparsing:       2.0.3-py27_0            
    pyqt:            4.11.4-py27_1           
    python:          2.7.11-0                
    python-dateutil: 2.4.2-py27_0            
    pytz:            2015.7-py27_0           
    qimage2ndarray:  1.4-py27_1              
    qt:              4.8.7-1                 
    rank_filter:     0.4.6-nppy27            
    readline:        6.2-2                   
    scikit-image:    0.11.3-np110py27_0      
    scipy:           0.16.1-py27_openblas_103 [openblas]
    setuptools:      19.6.2-py27_0           
    sip:             4.16.9-py27_0           
    six:             1.10.0-py27_0           
    spams:           v2.5-py27_openblas_10    [openblas]
    sqlite:          3.9.2-0                 
    tk:              8.5.18-0                
    vigra:           gca22467-py27_4         
    volumina:        v1.1.8-py27_1           
    wheel:           0.26.0-py27_1           
    xz:              5.0.5-0                 
    zlib:            1.2.8-0 
```
I think this behavior looks right, but I am curious to hear your thoughts.

**Comment by jakirkham** [comment]
Interesting. So, the reason I have a version of OpenBLAS that doesn't track features was as a workaround for an issue that resulted in no package being installed when a non-feature tracking version of a package was missing. I went ahead and removed the non-feature versions if you want to give it a try and see how goes.

**Comment by mcg1969** [comment]
The problem is that the defaults channel has the non-feature version of `openblas`, and currently when there are packages in two channels with the same name, they get effectively "merged". So you're going to have this problem as long as you insist on putting the `track_feature` inside of `openblas` itself.

**Comment by mcg1969** [comment]
I consider the way we handle name collisions across channels to be suboptimal, but it's really not something I intend to address yet. We have some larger ambitions for that.

**Comment by ukoethe** [comment]
In response to https://github.com/conda/conda/issues/1476#issuecomment-180611390
> My thinking is that if a feature is present in the environment, any package that has that among its options should use it, even if that means selecting an older version.

That is the behavior I intuitively expected during my first steps with conda, so I vote +1 for making this the default. On the other hand, users should have the option to override it, as the following use case demonstrates: Suppose a package comes in two variants, a pure Python one (without any features), and a faster compiled one (with the appropriate compiler feature). When the compiler feature is being tracked, `conda install` should pick the second variant by default, but give the user the option to choose the first. This is especially useful during development, when the newest version only exists in the pure Python form.

Another important consideration is to prevent recipe designers from accidentally forgetting a feature. I'm wondering if it makes sense to implement an `auto` keyword for features:
```
build:
  features: - auto
```
If `auto` is present, `conda build` would read all `track_features:` specifications from direct dependencies (specifically, from the precise variants of these dependencies in the build environment) and insert them automatically into the present recipe.

**Comment by mcg1969** [comment]
I don't think the `conda build auto` idea should be necessary. If a dependency has a `track_feature` then there should be no need to copy it.

**Comment by mcg1969** [comment]
As for overriding the default behavior, that can currently be done on a per-version basis by specifying the build string as well in the match spec. What might be nice is to specify the _feature list_ explicitly in the spec.

**Comment by ukoethe** [comment]
> If a dependency has a track_feature then there should be no need to copy it.

I probably didn't make myself clear. Suppose there is an openblas package whose `meta.yaml` reads
```
build:
  track_features: - openblas_0_2
```
Then a numpy recipe might specify:
```
requirements:
   build:
     - openblas
```
and conda-build choses to resolve the dependency into the openblas package above. The thus created numpy package should now contain `features: - openblas_0_2` because it is probably binary incompatible with other BLAS implementations. Instead of relying on numpy's recipe authors to add this `features` specification manually, the proposed `features: - auto` would take care of turning a `track-features: - a_feature` in the dependency into a `features: - a_feature` spec in the dependent automatically. Does this make sense?

**Comment by ukoethe** [comment]
> As for overriding the default behavior, that can currently be done on a per-version basis by specifying the build string as well in the match spec.

Yes, I've noticed the notation `openblas=0.2.15=101` in your example above. It would be nice to mention this elegant possibility in the documentation of the `conda install` command.

**Comment by mcg1969** [comment]
@ukoethe, I do see your point. I need to think about this a bit. (But this is actually a `conda-build` issue regardless). 

I think part of the problem here is that we need to define and document some sensible practices around the use of `track_features`. It seems problematic to me for some `openblas` packages to have a `track_features` value and for some not to. Heck, it seems problematic to me for packages to have multiple, different `track_features` values. We're currently doing that with `python` but I'm encouraging us to move away from that for various reasons.

So for example, what happens if you build a package with an `openblas` dependency and there are multiple `track_features` to choose from?

**Comment by mcg1969** [comment]
> Yes, I've noticed the notation openblas=0.2.15=101 in your example above. It would be nice to mention this elegant possibility in the documentation of the conda install command.

Good idea. I'm actually wondering if we shouldn't offer a richer, extensible match spec option, something that looks like Python keyword/value pairs; e.g., `openblas(version=0.2.15,build=101,feature=openblas)`

**Comment by ukoethe** [comment]
> Python keyword/value pairs; e.g., openblas(version=0.2.15,build=101,feature=openblas)

I love keyword arguments. However, using `(  )` and `,` as delimiters is problematic because these characters have special meaning in the shell. Characters like `/`, `?` or `@` are more suitable.

**Comment by ukoethe** [comment]
> I think part of the problem here is that we need to define and document some sensible practices around the use of `track_features`.

For starters, we should create a list of configuration options that we want to address with `track_features`. So far, I've seen three: the compiler, the Python version, and the BLAS variant. Anything else?

> It seems problematic to me for some openblas packages to have a track_features value and for some not to.

Hopefully, this will only happen for historical reasons. Once the `track_features` mechanism is fully functional, `anaconda upload` should probably issue a warning or even an error if one attempts to upload a package without `track_features` specification when all (or most or any) existing variants do have such a specification (along with suitable options to override the default policy when necessary).

> It seems problematic to me for packages to have multiple, different `track_features` values. We're currently doing that with python but I'm encouraging us to move away from that for various reasons.

In case of Python, the obvious solution is to introduce compiler runtime packages that contain the appropriate `track_features` spec (and install the corresponding libraries) and have Python depend on one of those. Thus, Python has a `features` spec for the compiler, and a `track_features` spec for its own version. I guess that splitting up a package into a core and a dependency will solve the problem of multiple `track_features` values in other cases as well.

> So for example, what happens if you build a package with an openblas dependency and there are multiple track_features to choose from?

I'm not sure if I understand the question. Anyway, I imagine (and implemented) the following workflow: to build a package against specific dependencies, one first creates and activates a 'bootstrap' environment that contains precisely the desired dependency variants. Then, `conda build` (possibly with the help of jinja2 templates and/or command line options) recreates the bootstrap environment in the `_build` environment and adds missing dependencies (if any) according to the standard version resolution rules. Placing a dependency in the bootstrap environment thus ensures that conda won't accidentally pick an undesired variant. Additional machinery (like my `auto` proposal) then ensures that the newly built package precisely records the variants it was compiled against, so that `conda install` is guaranteed to create a consistent installation. Keyword arguments to `conda install` allow users to easily customize the variants to be installed.

**Comment by jakirkham** [comment]
Sorry I am a bit behind, I will try to catch up on what is going on here soon.

I wanted to point out one related issue that I was seeing when using the `python setup.py bdist_conda` when using the `openblas` feature. This is all using `conda` 3.19.0 and `conda-build` 1.18.2 so is not using any of the changes to the solver, but I think this is a weird case to be aware of.

When building a package, it appears to do the right thing in the beginning. ( https://travis-ci.org/jakirkham/nanshe/builds/102645037#L262-L319 )

When doing the testing phase, it appears to pick up packages without the feature instead. ( https://travis-ci.org/jakirkham/nanshe/builds/102645037#L608-L678 )

This was one of the main motivators behind why I built packages with features and without as I wanted to make sure the linkages remained correct in cases where features were not respected.

I have seen some weird issues with `mkl` and `nomkl` along these lines. Namely some cases like this test phase of `python setup.py bdist_conda` where both `mkl` and `nomkl` will get installed.

**Comment by ukoethe** [comment]
> When doing the testing phase, it appears to pick up packages without the feature instead.

I noticed a similar (the same?) problem yesterday with conda-build 1.8.2: During the testing phase of my recipe, conda installed Python 3.5 although the run requirements explicitly stated that Python 2.7.* was needed. My impression is that this is not a problem with features, but that conda-build 1.8.2 ignores the run requirements entirely when setting up the test.

**Comment by mcg1969** [comment]
It's not clear to me how much of this is due to `conda` core and how much is due to `conda-build`, and we should try and tease that apart. Is there a reason you're not using conda-build 1.9.0, by the way?

**Comment by ukoethe** [comment]
conda-build 1.8.2 is what I got when I installed Miniconda3-latest two weeks ago. I (falsly?) concluded that it was still the recommended version.

**Comment by jakirkham** [comment]
> During the testing phase of my recipe, conda installed Python 3.5 although the run requirements explicitly stated that Python 2.7.* was needed.

@ukoethe, what python is in your `default` environment?

> It's not clear to me how much of this is due to `conda` core and how much is due to `conda-build`, and we should try and tease that apart.

Definitely, I'm a bit worried by the fact that the two stages do different things. It also makes me wonder if some code is being duplicate with some minor changes.

> Is there a reason you're not using conda-build 1.9.0, by the way?

Those builds were done a little while ago so that is why they were done with 1.18.2. Right now I'm struggling to keep `mkl` out of the environment so haven't had a successful build with 1.19.0 though I think it is this same bug manifesting in a different way. I will try installing the development version of `conda` in the CI and see if that helps today.

**Comment by jakirkham** [comment]
Ah, so I think that is the cause @ukoethe. Normally I have a copy of Miniconda 2 installed for `bdist_conda` when I want to build Python 2 packages and a copy of Miniconda 3 installed for building Python 3 packages. This is a very old issue actually. ( https://github.com/conda/conda-build/issues/132 )

**Comment by ukoethe** [comment]
@jakirkham Yes, I performed a build for Python 2.7 with Miniconda3. The build itself worked as it should (in particular, Python 2.7 was pulled into the `_build` environment, as stated in the build requirements), but the test failed because conda pulled Python 3.5 into the `_test` environment, contrary to what the run requirements said.

**Comment by ukoethe** [comment]
> It's not clear to me how much of this is due to conda core and how much is due to conda-build, and we should try and tease that apart.

I fully agree that conda core should be fixed first. The conda-build problems just keep getting into the way when I try to take advantage of your improvements.

**Comment by ukoethe** [comment]
I conducted tests with the new version resolution algorithm (as installed from github master) and found the following **bug**:

The active environment contained three mock-up packages:
* `test_msvc 11.0` containing `track_features: - vc11`
* `test_python 2,7` containing `track_features: - py27` and `features: - vc11`
* `test_openblas 0.2` containing `track_features: - openblas0_2` and `features: - vc11`

Then I called `conda install --use-local test_numpy` which produced the following output:
```
The following NEW packages will be INSTALLED:

    test_numpy:  1.11.0-vc11_py27_openblas0_2_0 [vc11 py27 openblas0_2]

The following packages will be UPDATED:

    test_python: 2.7-vc11_0 [vc11] --> 3.5-vc11_0 [vc11]
```
In other words, conda picked the desired version of `test_numpy`, but at the same time upgraded `test_python` to version 3.5, so that an inconsistent environment resulted. My desired behavior is that `test_python` remains untouched because the upgrade is not required to fulfill the installation request.

**Comment by ukoethe** [comment]
Regarding upgrades of packages in general, I'd prefer conda to upgrade only under two circumstances:
* the update is explicitly requested by the user (by providing an explicit version spec, a command line option, or a global configuration flag), or
* the present installation request cannot be fulfilled otherwise.

In any case, conda should refuse (with suitable diagnostic output) to perform the upgrade if it would result in an inconsistent environment.

**Comment by mcg1969** [comment]
I agree that the behavior you're describing is a bug. `conda` is *supposed* to add an implicit constraint to `python` to prevent it from being upgraded to Python 3 unless explicitly requested.

**Comment by mcg1969** [comment]
Oh wait, the problem is you're using `test_python`, not `python`. So the implicit constraint behavior of conda is not being activated here. Is there any chance you can share the `repodata.json` file of the local environment you created? That will be the easiest way for me to reproduce the issue and make sure `conda` is doing what we all expect.

**Comment by ukoethe** [comment]
I investigated the matter a bit further and found that the update occurs because the run requirements of `test_numpy` are
```
requirements:
  run:
    - test_python
    ...
```
If I change this into
```
requirements:
  run:
    - test_python 2.7
    ...
```
conda doesn't suggest to upgrade `test_python`. Once again, this is partly a conda-build problem: conda-build doesn't automatically specialize a general requirement like `- test_python` into what was actually present during the build (e.g. into `- test_python 2.7`).

I created a repository with my mock-up recipes at https://github.com/ukoethe/test_conda_solver. The problem occurs with `conda install test_numpy`, but not with `conda install test_numpy_exact` (which differ by their requirements as above). Here is my [repodata.json](https://github.com/conda/conda/files/125014/repodata.txt).
